src/backend/distributed/metadata/metadata_cache.c

```diff
@@ -4910,3 +4912,154 @@ poolinfo_valid(PG_FUNCTION_ARGS)
 							"with Citus, please contact us at "
 							"https://citusdata.com/about/contact_us")));
 }
+
+
+/*
+ * GetLocalClockValue returns logical_clock_value value stored in
+ * the Citus catalog pg_dist_local_group in given "savedValue"
+ * arg.
+ *
+ * Returns true on success, i.e.: finds a record in pg_dist_local_group,
+ * and false otherwise.
+ */
+bool
+GetLocalClockValue(uint64 *savedValue)
+{
+	ScanKey scanKey = NULL;
+	int scanKeyCount = 0;
+	bool success;
+
+	InitializeCaches();
+
+	Relation pgDistLocalGroupId = table_open(DistLocalGroupIdRelationId(),
+											 AccessShareLock);
+
+	SysScanDesc scanDescriptor = systable_beginscan(pgDistLocalGroupId,
+													InvalidOid, false,
+													NULL, scanKeyCount, scanKey);
+
+	TupleDesc tupleDescriptor = RelationGetDescr(pgDistLocalGroupId);
+
+	HeapTuple heapTuple = systable_getnext(scanDescriptor);
+
+	if (HeapTupleIsValid(heapTuple))
+	{
+		Datum datumArray[Natts_pg_dist_local_group];
+		bool isNullArray[Natts_pg_dist_local_group];
+
+		heap_deform_tuple(heapTuple, tupleDescriptor, datumArray, isNullArray);
+
+		if (isNullArray[Anum_pg_dist_local_logical_clock_value - 1])
+		{
+			ereport(ERROR, (errmsg("unexpected: got null clock value "
+								   "stored in pg_dist_local_group")));
+		}
+
+		*savedValue = DatumGetInt64(
+			datumArray[Anum_pg_dist_local_logical_clock_value - 1]);
+		success = true;
+	}
+	else
+	{
+		/*
+		 * Upgrade is happening. When upgrading postgres, pg_dist_local_group is
+		 * temporarily empty before citus_finish_pg_upgrade() finishes execution.
+		 */
+		success = false;
+	}
+
+	systable_endscan(scanDescriptor);
+	table_close(pgDistLocalGroupId, AccessShareLock);
+
+	return success;
+}
+
+
+/*
+ * PersistLocalClockValue gets invoked in two places, one in Citus maintenance
+ * daemon, and when a ever a backend exits, but that still leaves a time
+ * window where the clock value used may not be current in the catalog.
+ * TBD: May be store it during Checkpoint too?
+ */
+void
+PersistLocalClockValue(int code, Datum argUnused)
+{
+	/* It's a no-op if any below are true
+	 * Dist cache(s) are not set up yet or
+	 * In the midst of creating or altering extension or
+	 * If the version checks are disabled
+	 */
+	if (IsInitProcessingMode() || creating_extension ||
```

Comment:
minor: given that it's already clear what we're doing here, we can maybe drop this comment ?

But maybe we could comment "why" we're doing that.

Reply:
nit: please let's also drop the parenthesis around (!IsMaintainanceDaemonProcess())